python/ray/air/preprocessors/stateful_preprocessor.py

```diff
@@ -0,0 +1,120 @@
+    Logic must be defined to perform fitting on a Ray Dataset and transforming
+    Pandas DataFrames.
+
+    Example:
+
+    .. code-block:: python
+
+        from typing import Dict
+
+        import pandas as pd
+        import ray.data
+        from pandas import DataFrame
+        from ray.air.preprocessors import StatefulPreprocessor
+        from ray.data import Dataset
+        from ray.data.aggregate import Max
+
+        items = [
+            {"A": 1, "B": 10},
+            {"A": 2, "B": 20},
+            {"A": 3, "B": 30},
+        ]
+
+        ds = ray.data.from_items(items)
+
+
+        def get_max_a(ds: Dataset):
+            # Calculate max value for column A.
+            max_a = ds.aggregate(Max("A"))
+            # {'max(A)': 3}
+            return max_a
+
+
+        def subtract_max_a_from_a_and_add_max_a_to_b(df: DataFrame, stats: Dict):
+            # Subtract max A value from column A and subtract it from B.
+            max_a = stats["max(A)"]
+            df["A"] = df["A"] - max_a
+            df["B"] = df["B"] + max_a
+            return df
+
+
+        preprocessor = StatefulPreprocessor(
+            get_max_a,
+            subtract_max_a_from_a_and_add_max_a_to_b
+        )
+        preprocessor.fit(ds)
+        transformed_ds = preprocessor.transform(ds)
+
+        expected_items = [
+            {"A": -2, "B": 13},
+            {"A": -1, "B": 23},
+            {"A": 0, "B": 33},
+        ]
+        expected_ds = ray.data.from_items(expected_items)
+        assert transformed_ds.take(3) == expected_ds.take(3)
+
+        batch = pd.DataFrame(
+            {
+                "A": [5, 6],
+                "B": [10, 10]
+            }
+        )
+        transformed_batch = preprocessor.transform_batch(batch)
+        expected_batch = pd.DataFrame(
+            {
+                "A": [2, 3],
+                "B": [13, 13],
+            }
+        )
+        assert transformed_batch.equals(expected_batch)
+
+
+    Args:
+        fit_fn: A user defined function that computes state information about
+            a :class:`ray.data.Dataset` and returns it in a :class:`dict`.
+        transform_fn: A user defined function that takes in a
+            :class:`pandas.DataFrame` and the :class:`dict` computed from
+            ``fit_fn``, and returns a transformed :class:`pandas.DataFrame`.
+    """
+
+    _is_fittable = True
+
+    def __init__(
+        self,
+        fit_fn: Callable[[Dataset], Dict],
+        transform_fn: Callable[["pandas.DataFrame", Dict], "pandas.DataFrame"],
+    ):
+        self.fit_fn = fit_fn
+        self.transform_fn = transform_fn
+
+    def _fit(self, dataset: Dataset) -> "Preprocessor":
+        self.stats_ = self.fit_fn(dataset)
+        return self
+
+    def _transform_pandas(self, df: "pandas.DataFrame") -> "pandas.DataFrame":
+        return self.transform_fn(df, self.stats_)
+
+    def __repr__(self):
+        fit_fn_name = getattr(self.fit_fn, "__name__", self.fit_fn)
```

Comment:
Supernit: I'm assuming that this is to handle the case when an instantiated callable class is passed? If so, you could have a more targeted fallback with something like:
```suggestion
        try:
            fit_fn_name = self.fit_fn.__name__
        except AttributeError:
            try:
                fit_fn_name = self.fit_fn.__class__.__name__
            except AttributeError:
                fit_fn_name = self.fit_fn
```
Always going with the `str(self.fit_fn)` fallback is probably fine (difference between displaying `fit_fn=<__main__.Foo at 0x7f4dde4f4710>` vs `fit_fn=Foo`), so feel free to ignore this if you don't think that it is worth it.

Reply:
Ah this was meant to be for the case where `fit_fn` is a lambda function. `str(self.fit_fn)` seems reasonable to me!